designdocs/state_storage.md

```diff
@@ -0,0 +1,101 @@
+one option might be to store this type of data on a blockchain as long as it is required.
+The scope for how long some of these stateful data should exist is dependent on the
+type of data. For example, some of them might exist as long as an active channel exists between two
+communicating parties and for some others it might be directly dependent on the existence of a
+particular party.
+
+### DSNP Usecase
+**Frequency** is the first implementation of [DSNP](https://dsnp.org) and most of the state
+transitions in **DSNP** can be modeled via `Announcements`. For some of these state transitions we
+only care about the latest state, and currently the only way to achieve this via announcements is to
+have some kind of third party indexer to track the latest state for these types.
+If there was a way to be able to mutate the state of a storage, then we could update it to its
+latest version and potentially remove the need for an indexer.
+
+## Goals
+1. Allowing storage of stateful data with flexible schemas on chain.
+2. Data stored for any `MSA` should not have any effects like changing read/write access times or
+storage costs for any data stored for another `MSA`.
+3. Allowing high write throughput.
+4. Allowing high read throughput.
+5. Avoiding data races between consecutive updates of the same item.
+
+## Non-goals
+- Defining an economic incentive to restrict infinite growth of on-chain database.
+
+## Glossary
+* `DSNP`: Decentralized Social Networking Protocol
+* `MSA`: Message Source Account. A registered identifier with the MSA pallet.
+
+## Proposal
+Creating a double map of (`SchemaId` and `PageNumber`) as keys and `vec<u8>` as stored values.
+
+Using mentioned double-map will allow storing data using a variety of schemas using pagination.To
+be able to achieve goal number 2 we will need to create a child-tree for each `MSA` and store this
+double map under these child trees for each `MSA`.
+
+### Modes of Storage
+- **Batch**: In this mode we are storing each entity of that schemaId as an item inside an array.
+This is recommended for items that are relatively small and there are more than a few of them.
+- **Single**: In this mode we are storing each entity individually, and it is recommended for items
+that are relatively large and there are a few of them.
+
+![image](https://user-images.githubusercontent.com/9152501/211664660-1d4aa9ac-ed11-44eb-a8f4-66c972750ac3.jpg)
+
+The rationale behind separating these modes is based on two reasons:
+- _Performance_: Batch modes always need 1 on-chain DB_READ and 1 DB_WRITE but Single mode might be
+done with only 1 DB_WRITE.
+- _Composability_: Batch mode allows better composability due to on-chain read requirement and smaller
+size of each item.
+
+### Actions
+1. **Append Item**
+    - _Modes_: Batch
+    - _input_: Serialized data `vec<u8>`
+    - _Purpose_: Adds a new item to existing array
+1. **Remove Item**
```

Comment:
Do we need to worry about gaps?

If I have pages (page size=100) 1-10 full and remove item from page 2, index 47, how does the next append work? does it fill the hole? Do we need to defrag from time to time?

Reply:
I think since we are doing all this page calculation and management at client side. We can find about a page which is not full and try to pack it, so if our clients behave properly we should have packed pages as much as possible but we might need to have a defrag process once in a while in case some client misbehave.

Reply:
To extend this approach we can do it in 2 different ways if there are more than 1 option(page) to append to

1- Deterministic: Always choosing the page with smallest (or maybe largest) index
  - _tradeoffs_: since the result would always be deterministic across different clients. Instead of having possibility of  duplicate entities inside different pages coming from different clients we will have collisions and one of them will fail but It would cause more collisions.

2-Random: Randomly choose the page from valid options
  - _tradeoffs_: Less collision since the probability of two clients writing to the same page would be lower but more probability of having the same identical item being duplicated in more than one page.